src/charm.py

```diff
@@ -305,180 +142,70 @@ def _on_relation_departed(self, event: RelationDepartedEvent, charm_state: Charm
+        self.model.unit.status = ops.MaintenanceStatus("Configuring Synapse")
+        helpers.configure_and_start_services(self, charm_state, container)
+        if helpers.is_mjolnir_enabled(self, charm_state):
+            self._mjolnir.enable(charm_state)
         if self.unit.is_leader():
-            logger.debug("Adding signing key to secret: %s", signing_key)
-            secret = self.app.add_secret({"secret-signing-key": signing_key})
-            peer_relation[0].data[self.app].update(
-                {"secret-signing-id": typing.cast(str, secret.id)}
-            )
-
-    def get_signing_key(self) -> typing.Optional[str]:
-        """Get signing key from secret.
+            self._matrix_auth.update_matrix_auth_integration(charm_state)
+        self._set_unit_status()
 
-        Returns:
-            Signing key as string or None if not found.
-        """
-        peer_relation = self.model.relations[synapse.SYNAPSE_PEER_RELATION_NAME]
-        if not peer_relation:
-            logger.error(
-                "Failed to get signing key: no peer relation %s found",
-                synapse.SYNAPSE_PEER_RELATION_NAME,
-            )
-            return None
-
-        secret_id = peer_relation[0].data[self.app].get("secret-signing-id")
-        if secret_id:
-            try:
-                secret = self.model.get_secret(id=secret_id)
-                logging.debug(secret.get_content().get("secret-signing-key"))
-                return secret.get_content().get("secret-signing-key")
-            except (ops.model.SecretNotFoundError, ValueError, TypeError) as exc:
-                logger.exception("Failed to get secret id %s: %s", secret_id, str(exc))
-                del peer_relation[0].data[self.app]["secret-signing-id"]
-        return None
+    def _set_unit_status(self) -> None:
+        """Set unit status depending on Synapse and NGINX state."""
+        if isinstance(self.unit.status, ops.BlockedStatus):
+            # Preserve error state set elsewhere
+            return
 
-    @inject_charm_state
-    def _on_leader_elected(self, _: ops.HookEvent, charm_state: CharmState) -> None:
-        """Handle Synapse leader elected event.
+        container = self.unit.get_container(synapse.SYNAPSE_CONTAINER_NAME)
+        if not container.can_connect():
+            self.unit.status = ops.MaintenanceStatus("Waiting for Synapse pebble")
+            return
 
-        This event handler will reconcile Synapse configuration after the following
-        scenarios:
-        - When the charm is deployed so the leader will be the main unit.
-        - When the leader, for any reason, has changed so the leader unit will be the main.
-        Once the peer data (main_unit_id) is changed, other units will emit reconcile and be
-        properly configured.
+        if not self._is_service_running(container, synapse.SYNAPSE_SERVICE_NAME):
+            self.unit.status = ops.MaintenanceStatus("Waiting for Synapse")
+            return
```

Comment:
We rely on having a next event, but if there's no next event, we could stay in `Maintenance` for ever.
Usually what we do is calling synchronously a health endpoint of the service (or the one of pebble checks) repeating it with some backoff. In case of failure, error out (as the service can't start).